### PR TITLE
Add Local_common recipe - great basis for a shared hosting workflow

### PR DESCRIPTION
**Pull request by johnny-bit**
# Local_common recipe

This recipe is a re-implementation of `common.php` recipe, for running some of the tasks locally instead of remotelly. This recipe is especially useful if hosting provider that You're deploying to has some limitations of what can/can't be run. Best example is a *Shared Hosting* environment.

### Installing

```php
// deploy.php

require 'vendor/deployphp/recipes/recipes/local_common.php';
```

### Configuration options

The `local_common.php` recipe doesn't expose any configuration options. Instead uses same options as standard `common.php` recipe.

### Environmental variables

- **local_git_cache** - Accepts *boolean*. Set to true if You wish to speed up cloning and You use full local-release workflow. This uses git with `--reference` and `--dissociate` set in order to speed up cloning.
- **local_deploy_path** - Accepts *string* specifying from where should deploys be done. This is functionally identical with `deploy_path` of `common.php` recipe.
- **local_release_path** - Accepts *string* specifying from where should deploys be done. This is functionally identical with `release_path` of `common.php` recipe. If You use full local release scenario, You do **not** wish to set this, as it's relative to `local_deploy_path`.

```php
env('local_git_cache', true);
env('local_deploy_path', '/tmp/deployer');
```

### Tasks

- `deploy:local:prepare` - Prepares local dirs for deployment. Instead of failing when `local_deploy_path` does not exist - tries to create it.
- `deploy:local:release` - Prepare release directory
- `deploy:local:update_code` - Clones repository into local release directory
- `deploy:local:vendors` - Run composer locally in release directory
- `deploy:local:symlink` - Symlink atomicly to newest release
- `current:local` - Show current symlinked release
- `cleanup:local` - Cleanup local releases.

### Suggested Usage

Depending on what *actually* is possible on remote deploy target or if You want to test deployment locally, there's many ways to utilize this recipe.

#### Scenario #1 

You just want to clone locally and run composer locally. Everything else can be done remotelly.

This can be set as follows:

```php
// deploy.php 

env('local_release_path', '/tmp/my_application');

/**
 * Main task
 */
task('deploy', [
    'deploy:prepare',
    'deploy:release',
    'deploy:local:update_code',
    'deploy:local:vendors',
    'deploy:symlink',
    'cleanup',
])->desc('Deploy your project');

after('deploy:local:vendors', function() {
  upload(env('local_release_path'), env('release_path'));
})->desc('Upload local to remote');
```

#### Scenario #2 

You use full local release *and* full remote release. Plus You wish to take advantage of `rsync.php` recipe.

```php
env('rsync_src', function(){
  $local_src = env('local_release_path');
  if(is_callable($local_src)){
    $local_src = $local_src();
  }
  return $local_src;
});

task('deploy', [
    'deploy:local:prepare',
    'deploy:prepare',
    'deploy:local:release',
    'deploy:release',
    'deploy:rsync_warmup',
    'deploy:local:update_code',
    'rsync',
    'deploy:local:symlink',
    'deploy:symlink',
    'cleanup',
    'cleanup:local'
])->desc('Deploy your project');

```

### CODE REVIEW

**Comment by elfet** [comment]
Cool stuff! Really want to merge it. What you think @gordalina, @ZeeCoder, @ZhukV ?

**Comment by gordalina** [comment]
Cool stuff here @johnny-bit !

1st: i would call it just 'local', there's no really any need to have common in the name
2nd: please expose the tasks prepended by 'local:*', this way we know that they are from this recipe.

**Comment by johnny-bit** [comment]
Urm.. Ok?

So: change `local_common` to local in both files then change task names?

Should task names be like `deploy:local:prepare` to `local:prepare`?? or
`local:deploy:prepare`?

should I also change `deploy:rsync_warmup` in #10 ?

**Comment by gordalina** [comment]
they should be `local:something something`, and yes #10 should be updated as well

**Comment by gordalina** [comment]
and just rename the files is ok :)

**Comment by johnny-bit** [comment]
How does it look now?

-- 
Pozdrawiam,
Hubert Kowalski

**Comment by gordalina** [comment]
Looks good now

**Comment by johnny-bit** [comment]
And #10?

-- 
Pozdrawiam,
Hubert Kowalski

**Comment by gordalina** [comment]
#10 as well, just waiting for @elfet

File: README.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ require 'vendor/deployphp/recipes/recipes/cachetool.php';
 | ------    | ----                                                                      | -----
 | cachetool | [read](http://github.com/deployphp/recipes/blob/master/docs/cachetool.md) | `require 'vendor/deployphp/recipes/recipes/cachetool.php';`
 | newrelic  | [read](http://github.com/deployphp/recipes/blob/master/docs/newrelic.md)  | `require 'vendor/deployphp/recipes/recipes/newrelic.php';`
+| local     | [read](http://github.com/deployphp/recipes/blob/master/docs/local.md)     | `require 'vendor/deployphp/recipes/recipes/local.php';`
```
Comment (gordalina) [inline README.md]: Please move this after `cachetool` and before `newrelic` :) 
lets keep it in alphabetical order

**Comment by elfet** [comment]
Cool! :)
I think it's time to merge it :)

File: docs/local.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Local_common recipe
+
+This recipe is a re-implementation of `common.php` recipe, for running some of the tasks locally instead of remotelly. This recipe is especially useful if hosting provider that You're deploying to has some limitations of what can/can't be run. Best example is a *Shared Hosting* environment.
+
+### Installing
+
+```php
+// deploy.php
+
+require 'vendor/deployphp/recipes/recipes/local_common.php';
```
Comment (gordalina) [inline docs/local.md]: this is only 'local'

File: docs/local.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+### Configuration options
+
+The `local_common.php` recipe doesn't expose any configuration options. Instead uses same options as standard `common.php` recipe.
```
Comment (gordalina) [inline docs/local.md]: here as well

File: README.md (hunk 1 above)
Comment (johnny-bit) [inline README.md]: Done.

btw: rabit was after slack ;)

2015-06-16 17:53 GMT+02:00 Samuel Gordalina <notifications@github.com>:

> In README.md
> <https://github.com/deployphp/recipes/pull/12#discussion_r32535883>:
>
> > @@ -32,6 +32,7 @@ require 'vendor/deployphp/recipes/recipes/cachetool.php';
> >  | ------    | ----                                                                      | -----
> >  | cachetool | [read](http://github.com/deployphp/recipes/blob/master/docs/cachetool.md) | `require 'vendor/deployphp/recipes/recipes/cachetool.php';`
> >  | newrelic  | [read](http://github.com/deployphp/recipes/blob/master/docs/newrelic.md)  | `require 'vendor/deployphp/recipes/recipes/newrelic.php';`
> > +| local     | [read](http://github.com/deployphp/recipes/blob/master/docs/local.md)     | `require 'vendor/deployphp/recipes/recipes/local.php';`
>
> Please move this after cachetool and before newrelic :)
> lets keep it in alphabetical order
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deployphp/recipes/pull/12/files#r32535883>.
>



-- 
Pozdrawiam,
Hubert Kowalski

**Comment by gordalina** [comment]
just a couple of notes, then its merge time :)

**Comment by gordalina** [comment]
@johnny-bit this now does not merge properly, please rebase with master and fix any conflicts.
Also you can add yourself as an author within composer.json

**Comment by johnny-bit** [comment]
Fix'd :)

2015-06-16 18:05 GMT+02:00 Samuel Gordalina <notifications@github.com>:

> @johnny-bit <https://github.com/johnny-bit> this now does not merge
> properly, please rebase with master and fix any conflicts.
> Also you can add yourself as an author within composer.json
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deployphp/recipes/pull/12#issuecomment-112480945>.
>



-- 
Pozdrawiam,
Hubert Kowalski

**Comment by gordalina** [comment]
Awesome, thanks for your work :smile: :beers:

**Comment by gordalina** [comment]
tagged as `3.0.1`

**Pull request merged**